packages/alfa-style/src/property/combinator/any.ts

```diff
@@ -0,0 +1,44 @@
+import { Parser } from "@siteimprove/alfa-parser";
+import { Err, Result } from "@siteimprove/alfa-result";
+
+export function any<I, T, E, A extends [] = []>(
+  parser: Parser<I, T, E, A>,
+  ...rest: Array<Parser<I, T, E, A>>
+): Parser<I, Iterable<T>, E, A> {
+  const parsers = [parser, ...rest];
+
+  return (input, ...args) => {
+    const seen = parsers.map(() => false);
+    const values: Array<T> = [];
+
+    outer: while (true) {
```

Comment:
I think we can get rid of the `outer` loop (and the awful goto) simply be replacing `continue outer` with `i=0` (and maybe go all the way and replace the inner `for` by a real `while` in that case…)

Reply:
That won't work I'm afraid. We want to delay evaluating lines 37-41 until no results were found for any of the parsers. Every time a result is found, the process restarts by skipping the rest of the outer loop.

This is another piece of code that could be improved by some kind of "reduce with the ability to break early" as we could then reduce the array of parsers to either an error or the _first_ result, after which we'd break the reduction. For lack of that, I'm fine with some local side effects and a labelled continue.

Reply:
🤔 but lines 37-41 would be after that loop…
```javascript
let i=0;
while (i<n) {
  if (seen[i]) continue; // skip parsers for components seen during a previous pass

  const result = parser[i](input); // try a new parser

  if (result.isErr()) { // parser is failing
    error = result;
    i++; // move on to next parser, within the same pass.
  } else {
    const [remainder, value] = result.get();
    values.push(value);
    input = remainder;
    seen[i] = true; // don't redo that parser during further passes.

    i = 0; // restarts a new pass from the start of the list of parsers.
  }
}
// we're only exiting the while if, during the last pass, each parser has either been seen or failed.

return values.length === 0 ? 
  error! // no parser has ever succeeded, returning last error found.
  : Result.of([input, values]); // some parsers have succeeded, returning result.

```

Reply:
Ah, now I see! That would indeed work 👍